cocotb/runner.py

```diff
@@ -366,99 +475,91 @@ def test_command(self) -> List[Command]:
+            cmds = [
+                ["iverilog", "-o", str(self.sim_file), "-D", "COCOTB_SIM=1", "-g2012"]
                 + self.get_define_options(self.defines)
                 + self.get_include_options(self.includes)
                 + self.get_parameter_options(self.parameters)
-                + self.compile_args
-                + self.verilog_sources
+                + self.build_args
+                + [str(source_file) for source_file in self.verilog_sources]
             ]
 
         else:
-            print("WARNING: Skipping compilation:" + self.sim_file)
+            print("WARNING: Skipping compilation of", self.sim_file)
 
-        return cmd
+        return cmds
 
 
 class Questa(Simulator):
+    supported_gpi_interfaces = {"verilog": ["vpi"], "vhdl": ["fli", "vhpi"]}
+
     @staticmethod
     def simulator_in_path() -> None:
         if shutil.which("vsim") is None:
             raise SystemExit("ERROR: vsim executable not found!")
 
-    def check_toplevel_lang(self, toplevel_lang: Optional[str]) -> str:
-        if toplevel_lang is None:
-            if self.vhdl_sources and not self.verilog_sources:
-                return "vhdl"
-            elif self.verilog_sources and not self.vhdl_sources:
-                return "verilog"
-            else:
-                raise ValueError("Must specify a toplevel_lang in a mixed design")
-        elif toplevel_lang in ("verilog", "vhdl"):
-            return toplevel_lang
-        else:
-            raise ValueError(
-                f"Questa does not support {toplevel_lang!r} as a toplevel_lang"
-            )
-
     @staticmethod
-    def get_include_options(includes: Sequence[str]) -> List[str]:
-        return ["+incdir+" + as_tcl_value(dir) for dir in includes]
+    def get_include_options(includes: Sequence[PathLike]) -> Command:
+        return [f"+incdir+{as_tcl_value(str(include))}" for include in includes]
 
     @staticmethod
-    def get_define_options(defines: Sequence[str]) -> List[str]:
-        return ["+define+" + as_tcl_value(define) for define in defines]
+    def get_define_options(defines: Mapping[str, object]) -> Command:
+        return [
+            f"+define+{as_tcl_value(name)}={as_tcl_value(str(value))}"
+            for name, value in defines.items()
+        ]
 
     @staticmethod
-    def get_parameter_options(parameters: Mapping[str, object]) -> List[str]:
-        return ["-g" + name + "=" + str(value) for name, value in parameters.items()]
+    def get_parameter_options(parameters: Mapping[str, object]) -> Command:
+        return [f"-g{name}={value}" for name, value in parameters.items()]
 
     def build_command(self) -> List[Command]:
 
-        cmd = []
+        cmds = []
 
         if self.vhdl_sources:
-            cmd.append(["vlib", as_tcl_value(self.library_name)])
-            cmd.append(
+            cmds.append(["vlib", as_tcl_value(self.hdl_library)])
+            cmds.append(
                 ["vcom", "-mixedsvvh"]
```

Comment:
Not everyone has licenses for mixedsvvh mode. Also there are VHDL 2008 features that work in normal mode that don't work in mixedsvvh mode. This flag should only be added if a library has mixed sources.

Also this shows the issues with separate VHDL_SOURCES and VERILOG_SOURES. In mixed language designs the order of VHDL vs VERILOG sources matters, so they should not be considered separately.

Reply:
>Not everyone has licenses for mixedsvvh mode.

Addressed in https://github.com/cocotb/cocotb/pull/3103

>Also this shows the issues with separate VHDL_SOURCES and VERILOG_SOURES. In mixed language designs the order of VHDL vs VERILOG sources matters, so they should not be considered separately.

This should be possible. Multiple `build()` calls in given order. Do you have something else in mind.